rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,881 @@
+
+
+        // Update stable memory with new association.
+        let assoc = Some((RegionId::from_id((*r).id), i as u16));
+        meta_data::block_region_table::set(BlockId(block_id), assoc.clone());
+
+
+        new_pages.set_ith_block_id(i, &BlockId(block_id));
+
+
+
+    write_with_barrier(mem, &mut (*r).vec_pages, new_vec_pages);
+    old_page_count.into()
+}
+
+pub(crate) unsafe fn region_load<M: Memory>(_mem: &mut M, r: Value, offset: u64, dst: &mut [u8]) {
+    use crate::ic0_stable::nicer::read;
+    use meta_data::size::BLOCK_IN_BYTES;
+
+    let r = RegionObject::from_value(&r);
+    if dst.len() > 1 {
+        r.check_relative_into_absolute_offset(offset + dst.len() as u64 - 1);
+    } else {
+        r.check_relative_into_absolute_offset(offset);
+    }
+    let (b1_off, b1, b1_len, b2) = r.relative_into_absolute_span(offset, dst.len() as u64);
+    if b1 == b2 {
+        // Case: only uses a single block, and thus only requires one read.
+        read(b1_off, dst);
+    } else {
+        // Case: Staged reads, one per block that holds requested data.
+
+        let mut i = 0; // invariant: i = # of bytes loaded.
+        let mut s = b1_off; // source of bytes, as absolute index.
+        let mut d = dst.as_mut_ptr(); // dest for bytes.
+
+        // do initial read (a special case, generally not full block length).
+        read(s, core::slice::from_raw_parts_mut(d, b1_len as usize));
+
+        // Advance input and output positions (i, d and s respectively).
+        i += b1_len;
+        d = d.offset(b1_len as isize);
+
+        // Do rest of block-sized reads.
+        // (invariant: they always occur at the start of a block).
+        loop {
+            if false {
+                println!(80, "load r={:?} s={} i={}", r.id(), s, i);
+            }
+            let (s_, _, b_len) = r.relative_into_absolute_info(offset + i);
+            s = s_;
+            if i + b_len > dst.len() as u64 {
+                // case: last (generally partial) block.
+                if dst.len() as u64 > i {
+                    read(
+                        s,
+                        core::slice::from_raw_parts_mut(d, dst.len() - i as usize),
+                    );
+                }
+                break;
+            } else {
+                // case: internal (full) block.
+                read(
+                    s,
+                    core::slice::from_raw_parts_mut(d, BLOCK_IN_BYTES as usize),
+                );
+                d = d.offset(BLOCK_IN_BYTES as isize);
+                i += BLOCK_IN_BYTES;
+            }
+        }
+    }
+}
+
+pub(crate) unsafe fn region_store<M: Memory>(_mem: &mut M, r: Value, offset: u64, src: &[u8]) {
+    use crate::ic0_stable::nicer::write;
+    use meta_data::size::BLOCK_IN_BYTES;
+
+    let r = RegionObject::from_value(&r);
+    if src.len() > 1 {
+        r.check_relative_into_absolute_offset(offset + src.len() as u64 - 1);
```

Comment:
Again, I think this addition can overflow in production Rust and incorrectly pass the check.